File: built-in plugins dsp/mb-restore-mouth-declick.cpp
/**
 * MB Mouth De-Click
 * Category : effect
 * Type     : gate
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Specialized mouth click and lip smack removal
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_RESTORE_MOUTH_DECLICK_H
#define MB_RESTORE_MOUTH_DECLICK_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbRestoreMouthDeclick : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-restore-mouth-declick";
    static constexpr const char* PLUGIN_NAME    = "MB Mouth De-Click";
    static constexpr const char* PLUGIN_TYPE    = "gate";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float sensitivity = 0.5f;  // range [0, 1]
    float clickLength = 5f;  // range [1, 20]
    float mode = 0f;  // range [0, 2]
    };

    MbRestoreMouthDeclick() = default;
    ~MbRestoreMouthDeclick() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.sensitivity = std::clamp(params.sensitivity, 0f, 1f);
        params.clickLength = std::clamp(params.clickLength, 1f, 20f);
        params.mode = std::clamp(params.mode, 0f, 2f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Mouth De-Click
        return input;
    }
};

#endif // MB_RESTORE_MOUTH_DECLICK_H

File: built-in plugins dsp/mb-restore-mouth-declick.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbRestoreMouthDeclickPlugin: PluginDefinition = { id: 'mb-restore-mouth-declick', slug: 'mb-restore-mouth-declick', name: 'MB Mouth De-Click', category: 'effect', type: 'gate' as any, version: '1.0.0', description: 'Specialized mouth click and lip smack removal', author: 'Max Booster', grade: 'A', parameters: [{ id: 'sensitivity', name: 'Sensitivity', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'clickLength', name: 'Click Length', type: 'float', defaultValue: 5, minValue: 1, maxValue: 20, automatable: false }, { id: 'mode', name: 'Mode', type: 'float', defaultValue: 0, minValue: 0, maxValue: 2, automatable: false }], defaultPreset: { sensitivity: 0.5, clickLength: 5, mode: 0 } };

export default MbRestoreMouthDeclickPlugin;
